File: backend/app/services/inventario_service.py
from datetime import datetime

from sqlalchemy.orm import Session

from app.models.movimiento_inventario import MovimientoInventario, TipoMovimiento
from app.repositories import movimiento_inventario_repository, producto_repository, stock_sucursal_repository
from app.services import auditoria_service, reorden_service


class ProductoNoEncontradoError(Exception):
    pass


class StockInsuficienteError(Exception):
    pass


def listar_movimientos(
    db: Session,
    sucursal_id: int,
    producto_id: int | None,
    q: str | None,
    tipo: TipoMovimiento | None,
    desde: datetime | None,
    hasta: datetime | None,
    page: int,
    size: int,
) -> tuple[list[MovimientoInventario], int]:
    return movimiento_inventario_repository.get_all(db, sucursal_id, producto_id, q, tipo, desde, hasta, page, size)


def registrar_movimiento(
    db: Session,
    usuario_id: int,
    producto_id: int,
    sucursal_id: int,
    tipo: TipoMovimiento,
    cantidad: int,
    motivo: str | None = None,
) -> MovimientoInventario:
    if producto_repository.get_by_id(db, producto_id) is None:
        raise ProductoNoEncontradoError(producto_id)

    # el stock ya no vive en Producto: el lock es sobre la fila de stock_sucursal de esta sucursal
    stock = stock_sucursal_repository.get_or_create_for_update(db, producto_id, sucursal_id)
    if tipo == TipoMovimiento.SALIDA and cantidad > stock.cantidad:
        raise StockInsuficienteError(producto_id)

    stock.cantidad += cantidad if tipo == TipoMovimiento.ENTRADA else -cantidad
    stock_sucursal_repository.save(db, stock)

    movimiento = MovimientoInventario(
        producto_id=producto_id,
        sucursal_id=sucursal_id,
        tipo=tipo,
        cantidad=cantidad,
        stock_resultante=stock.cantidad,
        motivo=motivo,
        usuario_id=usuario_id,
    )
    movimiento = movimiento_inventario_repository.create(db, movimiento)
    auditoria_service.registrar(
        db,
        usuario_id,
        "movimiento_inventario_registrado",
        "movimiento_inventario",
        movimiento.id,
        {"producto_id": producto_id, "tipo": tipo.value, "cantidad": cantidad, "stock_resultante": stock.cantidad},
    )

    if tipo == TipoMovimiento.SALIDA:
        reorden_service.disparar_si_corresponde(db, producto_id, sucursal_id)

    return movimiento

File: frontend/src/services/inventarioService.ts
import { api } from '@/services/api'
import type { Producto } from '@/services/productoService'
import type { PaginatedResponse } from '@/services/pagination'

export type TipoMovimiento = 'entrada' | 'salida'

export interface Movimiento {
  id: number
  producto_id: number
  producto: Producto
  tipo: TipoMovimiento
  cantidad: number
  stock_resultante: number
  motivo: string | null
  created_at: string
}

export interface MovimientoPayload {
  producto_id: number
  tipo: TipoMovimiento
  cantidad: number
  motivo: string | null
}

export interface ListMovimientosParams {
  productoId?: number
  q?: string
  tipo?: TipoMovimiento
  desde?: string
  hasta?: string
  sucursalId?: number | null
  page?: number
  size?: number
}

export async function listMovimientos(params: ListMovimientosParams = {}): Promise<PaginatedResponse<Movimiento>> {
  const { productoId, q, tipo, desde, hasta, sucursalId, page, size } = params
  const { data } = await api.get<PaginatedResponse<Movimiento>>('/inventario/movimientos', {
    params: {
      producto_id: productoId,
      q: q || undefined,
      tipo: tipo || undefined,
      desde: desde || undefined,
      hasta: hasta || undefined,
      sucursal_id: sucursalId ?? undefined,
      page,
      size,
    },
  })
  return data
}

export async function createMovimiento(payload: MovimientoPayload, sucursalId?: number | null): Promise<Movimiento> {
  const { data } = await api.post<Movimiento>('/inventario/movimientos', payload, {
    params: { sucursal_id: sucursalId ?? undefined },
  })
  return data
}
